crates/fiber-lib/src/fiber/channel.rs

```diff
@@ -832,54 +846,167 @@ where
                 }
-                CkbInvoiceStatus::Paid => {
-                    // we have already checked invoice status in apply_add_tlc_operation_with_peeled_onion_packet
-                    // this maybe happened when process is killed and restart
-                    error!("invoice already paid, ignore");
+
+                if tlcs.iter().any(|t| t.total_amount != tlc.total_amount) {
+                    return TlcSettleDownResult::Failed(
+                        tlcs,
+                        RemoveTlcReason::RemoveTlcFail(TlcErrPacket::new(
+                            TlcErr::new(TlcErrorCode::IncorrectOrUnknownPaymentDetails),
+                            &tlc.shared_secret,
+                        )),
+                    );
                 }
-                _ => {
-                    // do nothing
-                    // invoice status will be updated to paid after apply remove tlc operation
+
+                let total_tlc_amount = tlcs.iter().map(|tlc| tlc.amount).sum::<u128>();
+                let is_fulfilled = total_tlc_amount >= total_amount;
+                if invoice.allow_mpp() {
+                    // check payment secret for MPP payment
+                    let payment_secret = invoice.payment_secret();
+                    if payment_secret.is_none() || tlc.payment_secret.as_ref() != payment_secret {
+                        return TlcSettleDownResult::Failed(
+                            tlcs,
+                            RemoveTlcReason::RemoveTlcFail(TlcErrPacket::new(
+                                TlcErr::new(TlcErrorCode::IncorrectOrUnknownPaymentDetails),
+                                &tlc.shared_secret,
+                            )),
+                        );
+                    }
+
+                    // hold the tlc if support MPP and invoice is not fulfilled
+                    if !is_fulfilled {
+                        self.store.insert_hold_tlc(
+                            tlc.payment_hash,
+                            HoldTlc {
+                                channel_id: state.get_id(),
+                                tlc_id: tlc.tlc_id.into(),
+                                hold_expire_at: now_timestamp_as_millis_u64()
+                                    + DEFAULT_HOLD_TLC_TIMEOUT,
+                            },
+                        );
+                        return TlcSettleDownResult::Hold;
+                    }
+                } else {
+                    // single path payment
+                    // fail tlc if invoice is not fulfilled
+                    if !is_fulfilled {
+                        return TlcSettleDownResult::Failed(
+                            tlcs,
+                            RemoveTlcReason::RemoveTlcFail(TlcErrPacket::new(
+                                TlcErr::new(TlcErrorCode::AmountBelowMinimum),
+                                &tlc.shared_secret,
+                            )),
+                        );
+                    }
```

Comment:
this will never reached for non MPP payment? since we have check before this on line: 899.